### PR TITLE
refactor(vue): revised menu & tests

### PR DESCRIPTION
**Pull request by iamdin**
- [x] revised menu components
- [x] stories
- [x] tests

### CODE REVIEW

**Comment by Omikorin** [comment]
Tell me when you finish :)

**Review by iamdin — COMMENTED** [review]
> Tell me when you finish :) 

Hi, i thinks it's ready for review. Thanks.

File: packages/frameworks/vue/src/menu/menu-item.tsx [hunk 1]
```diff
@@ -5,7 +5,7 @@ import { type Assign } from '../types'
 import { type ComponentWithProps } from '../utils'
 import { useMenuContext } from './menu-context'
 
-export type MenuItemProps = Assign<HTMLArkProps<'button'>, ItemProps>
+export interface MenuItemProps extends Assign<HTMLArkProps<'button'>, ItemProps> {}
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu-item.tsx:8]:
```suggestion
export interface MenuItemProps extends Assign<HTMLArkProps<'div'>, ItemProps> {}
```

File: packages/frameworks/vue/src/menu/menu-content.tsx [hunk 2]
```diff
@@ -2,11 +2,13 @@ import { defineComponent } from 'vue'
...
 import { emits, props } from '../presence/presence.props'
+import type { Assign } from '../types'
 import { useMenuContext } from './menu-context'
+import type { ComponentWithProps } from '../utils'
 
-export type MenuContentProps = HTMLArkProps<'div'> & PresenceProps
+export interface MenuContentProps extends Assign<HTMLArkProps<'div'>, PresenceProps> {}
 
-export const MenuContent = defineComponent({
+export const MenuContent: ComponentWithProps<MenuContentProps> = defineComponent({
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu-content.tsx:11]:
```suggestion
export const MenuContent = defineComponent({
```
We don't use `ComponentWithProps<P>` lately

File: packages/frameworks/vue/src/menu/menu-trigger.tsx [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
 export const MenuTrigger = defineComponent({
   name: 'MenuTrigger',
   setup(_, { slots, attrs }) {
     const api = useMenuContext()
-    return () => {
-      const DefaultSlot = useUniqueChild(slots, 'MenuTrigger')
-
-      return h(DefaultSlot, { ...api.value.triggerProps, ...attrs })
-    }
+
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu-trigger.tsx:11]:
```suggestion

```

ESLint found spaces

File: packages/frameworks/vue/src/menu/tests/context-menu.test.vue [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<script lang="ts" setup>
+import { Menu } from '..'
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/tests/context-menu.test.vue:2]:
```suggestion
import { Menu } from '../'
```

File: packages/frameworks/vue/src/menu/tests/menu-item-group.test.vue [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<script lang="ts" setup>
+import { Menu } from '..'
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/tests/menu-item-group.test.vue:2]:
```suggestion
import { Menu } from '../'
```

File: packages/frameworks/vue/src/menu/tests/menu.test.vue [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<script lang="ts" setup>
+import { Menu } from '..'
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/tests/menu.test.vue:2]:
```suggestion
import { Menu } from '../'
```

File: packages/frameworks/vue/src/menu/tests/nested-menu.test.vue [hunk 7]
```diff
@@ -0,0 +1,25 @@
+<script lang="ts" setup>
+import { Menu } from '..'
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/tests/nested-menu.test.vue:2]:
```suggestion
import { Menu } from '../'
```

File: packages/frameworks/vue/src/menu/tests/option-groups.test.vue [hunk 8]
```diff
@@ -0,0 +1,46 @@
+<script lang="ts" setup>
+import { ref } from 'vue'
+import { Menu } from '..'
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/tests/option-groups.test.vue:3]:
```suggestion
import { Menu } from '../'
```

File: packages/frameworks/vue/src/menu/use-menu.ts [hunk 9]
```diff
@@ -1,43 +1,63 @@
...
+  const [state, send, machine] = useMachine(
     menu.machine({
-      ...reactiveContext,
-      id: reactiveContext.id || useId().value,
+      ...context.value,
+      id: context.value.id || useId().value,
       getRootNode,
-      onOpenChange() {
-        emit('open-change')
+      onOpenChange(details) {
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/use-menu.ts:32]:
```suggestion
      onOpenChange: (details) => {
```
I hope it doesn't break anything. We use arrow functions for emitters.

File: packages/frameworks/vue/src/menu/use-menu.ts [hunk 10]
```diff
@@ -1,43 +1,63 @@
...
+        emit('open-change', details)
       },
-      onSelect() {
-        emit('select')
+      onSelect(details) {
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/use-menu.ts:35]:
```suggestion
      onSelect: (details) => {
```

File: packages/frameworks/vue/src/menu/use-menu.ts [hunk 11]
```diff
@@ -1,43 +1,63 @@
...
+        emit('select', details)
       },
-      onValueChange() {
-        emit('value-change')
-        emit('update:modelValue')
+      onValueChange(details) {
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/use-menu.ts:38]:
```suggestion
      onValueChange: (details) => {
```

**Review by Omikorin — CHANGES_REQUESTED** [review]
I took a quick look. First of all, great work! I'll test everything tomorrow on my PC.

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+        <Menu.Positioner>
+          <Menu.Content lazy-mount>
+            <Menu.Item id="search">Search</Menu.Item>
+            <Menu.Item id="undo">Undo</Menu.Item>
+            <Menu.Item id="delivery" disabled> Delivery </Menu.Item>
+            <Menu.Item id="unlink">Unlink</Menu.Item>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:27]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+        <Menu.Positioner>
+          <Menu.Content>
+            <Menu.Item id="search">Search</Menu.Item>
+            <Menu.Item id="undo">Undo</Menu.Item>
+            <Menu.Item id="delivery" disabled> Delivery </Menu.Item>
+            <Menu.Item id="unlink">Unlink</Menu.Item>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:41]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+            </Menu.ItemGroup>
+            <Menu.ItemGroup id="group-2">
+              <Menu.ItemGroupLabel htmlFor="group-2">Group 2</Menu.ItemGroupLabel>
+              <Menu.Item id="rename">Rename...</Menu.Item>
+              <Menu.Item id="delete">Delete...</Menu.Item>
+            </Menu.ItemGroup>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:60]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+          <Menu.Content>
+            <Menu.Item id="search">Search</Menu.Item>
+            <Menu.Item id="undo">Undo</Menu.Item>
+            <Menu.Separator />
+            <Menu.Item id="delivery" disabled> Delivery </Menu.Item>
+            <Menu.Item id="unlink">Unlink</Menu.Item>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:74]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+        <Menu.Positioner>
+          <Menu.Content>
+            <Menu.Item id="search">Search</Menu.Item>
+            <Menu.Item id="undo">Undo</Menu.Item>
+            <Menu.Item id="delivery" disabled> Delivery </Menu.Item>
+            <Menu.Item id="unlink">Unlink</Menu.Item>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:91]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+                <Menu.Content>
+                  <Menu.Item id="twitter">Twitter</Menu.Item>
+                  <Menu.Item id="message">Message</Menu.Item>
+                </Menu.Content>
+              </Menu.Positioner>
+            </Menu.Root>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:112]:
```suggestion
    </Variant>

```

File: packages/frameworks/vue/src/menu/menu.story.vue [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+                value="chakra"
+                v-slot="{ isChecked }"
+              >
+                {{ isChecked ? '✅' : '' }} chakra
+              </Menu.OptionItem>
+            </Menu.ItemGroup>
+          </Menu.Content>
+        </Menu.Positioner>
+      </Menu.Root>
+    </Variant>
```
Comment (Omikorin) [inline packages/frameworks/vue/src/menu/menu.story.vue:163]:
```suggestion
    </Variant>

```

**Review by Omikorin — CHANGES_REQUESTED** [review]

File: packages/frameworks/vue/src/menu/menu-item.tsx (hunk 1 above)
Comment (iamdin) [reply packages/frameworks/vue/src/menu/menu-item.tsx]: good catch, i missed it

File: packages/frameworks/vue/src/menu/menu-content.tsx (hunk 2 above)
Comment (iamdin) [reply packages/frameworks/vue/src/menu/menu-content.tsx:11]: yes, i think so. `ComponentWithProps<P>` has a little hack. I think we just need define props correctly with `defineComponent`, or use `PropType` to annotate complex props types.

**Comment by iamdin** [comment]
> I took a quick look. First of all, great work! I'll test everything tomorrow on my PC.

Thanks! Feel great to contribute to ark-ui. Wating for your test.
And i will continue to contribute to ark. Looking forward to vuejs 1.0.

File: packages/frameworks/vue/src/menu/menu-content.tsx (hunk 2 above)
Comment (iamdin) [reply packages/frameworks/vue/src/menu/menu-content.tsx:11]: > We don't use `ComponentWithProps<P>` lately. We could reconsider this. I found some little differences in intellisense, but in any way, this works. What do you think @iamdin?
> 
> Edit: According to #1884, it should be better to avoid this approach. Please remove each occurrence of `ComponentWithProps`编辑：根据#1884，最好避免这种方法。请删除所有出现的 `ComponentWithProps`

Yep, I will refactor with the new feature [Function Signature](https://vuejs.org/api/general.html#function-signature) released in Vue3.3

Comment (Omikorin) [reply packages/frameworks/vue/src/menu/menu-content.tsx]: Wow, they simplified a case for generic types in components, nice.

Comment (iamdin) [reply packages/frameworks/vue/src/menu/menu-content.tsx]: Finally,  just remove `ComponentWithProps`

File: packages/frameworks/vue/src/factory.tsx [hunk 19]
```diff
@@ -65,7 +65,7 @@ export function withAsChild(__component: RenderFunctionArgs) {
     },
     setup(props, { attrs, slots }) {
       const instance = getCurrentInstance()
-      if (!props.asChild) return () => <__component {...attrs}>{slots.default?.()}</__component>
+      if (!props.asChild) return () => h(__component, { ...attrs }, slots.default?.())
```
Comment (Omikorin) [inline packages/frameworks/vue/src/factory.tsx:68]: Why this change?

Comment (iamdin) [reply packages/frameworks/vue/src/factory.tsx:68]: fix a type error
image

Comment (Omikorin) [reply packages/frameworks/vue/src/factory.tsx:68]: Okay, it works and typecheck is good, so I approve this

**Review by Omikorin — APPROVED** [review]

**Comment by iamdin** [comment]
@Omikorin That's okay! Thanks for your time to review. Hope you can have time to rest.
